aptos-move/framework/aptos-framework/sources/IterableTable.move

```diff
@@ -0,0 +1,184 @@
+        table.head
+    }
+
+    /// Returns the key of the tail for iteration.
+    public fun tail_key<K: copy + store + drop, V: store>(table: &IterableTable<K, V>): Option<K> {
+        table.tail
+    }
+
+    /// Acquire an immutable reference to the IterableValue which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow_iter<K: copy + store + drop, V: store>(table: &IterableTable<K, V>, key: &K): (&V, Option<K>, Option<K>) {
+        let v = Table::borrow(&table.inner, key);
+        (&v.val, v.prev, v.next)
+    }
+
+    /// Acquire an immutable reference to the value and previous/next key which `key` maps to
+    /// Aborts if there is no entry for `key`.
+    public fun borrow_iter_mut<K: copy + store + drop, V: store>(table: &mut IterableTable<K, V>, key: &K): (&mut V, Option<K>, Option<K>) {
+        let v = Table::borrow_mut(&mut table.inner, key);
+        (&mut v.val, v.prev, v.next)
+    }
+
+    /// Remove from `table` and return the value and previous/next key which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun remove_iter<K: copy + store + drop, V: store>(table: &mut IterableTable<K, V>, key: &K): (V, Option<K>, Option<K>) {
+        let val = Table::remove(&mut table.inner, key);
+        if (Option::contains(&table.tail, key)) {
+            table.tail = val.prev;
+        };
+        if (Option::contains(&table.head, key)) {
+            table.head = val.next;
+        };
+        if (Option::is_some(&val.prev)) {
+            let key = Option::borrow(&val.prev);
+            Table::borrow_mut(&mut table.inner, key).next = val.next;
+        };
+        if (Option::is_some(&val.next)) {
+            let key = Option::borrow(&val.next);
+            Table::borrow_mut(&mut table.inner, key).prev = val.prev;
+        };
+        let IterableValue {val, prev, next} = val;
+        (val, prev, next)
+    }
+
+    /// Remove all items from v2 and append to v1.
+    public fun append<K: copy + store + drop, V: store>(v1: &mut IterableTable<K, V>, v2: &mut IterableTable<K, V>) {
+        let key = head_key(v2);
+        while (Option::is_some(&key)) {
+            let (val, _, next) = remove_iter(v2, Option::borrow(&key));
+            add(v1, Option::borrow(&key), val);
+            key = next;
+        };
+    }
+
+    #[test]
+    fun iterable_table_test() {
+        let table = new();
+        let i = 0;
+        while (i < 100) {
+            add(&mut table, &i, i);
+            i = i + 1;
+        };
+        assert!(length(&table) == 100, 0);
+        i = 0;
+        while (i < 100) {
+            assert!(remove(&mut table, &i) == i, 0);
+            i = i + 2;
+        };
+        assert!(!empty(&table), 0);
+        let key = head_key(&table);
+        i = 1;
+        while (Option::is_some(&key)) {
+            let (val, _, next) = borrow_iter(&table, Option::borrow(&key));
```

Comment:
You may not want expose `prev` and `next` to user to loop, IMHO.
My gut feeling about the API is:
```
struct Iterator {
    cur: Option<K>;
}

public fun IterableTable::iter(t: &IteratableTble): Iterator{
    Iterator {
        cur: head_hey(t);
    }
}
public fun Iterator::next(t: &mut IterableTable, i: &mut Iterator): &V {
    // if not none
    let v = Table::borrow(&table.inner, t.cur);
    t.cur = v.next; 
    &v.val
}

Reply:
fancy

Reply:
I can't make the fancy stuff work lol, the borrow checker is dumb. if you pass &t.cur to get the val, you can't mutate t.cur

Reply:
shit, the same problem in https://github.com/move-language/move/issues/95 
but I think you can copy `cur` in this case as a workaround?

Reply:
oh let me try it, such an advanced move programmer

Reply:
not working :) even after copy, the key is still a local variable that's dropped after the function end (shorter lifetime than the return value